Remove global spinlock for EH stacktrace

janvorli opened the pull request:
The global spinlock that was used to ensure that stack trace and the associated dynamic methods array were updated and read atomically. However, for the new EH, it has shown to cause a high contention in case many threads were handling exceptions at the same time.

This change replaces the two arrays by one object member in the exception class. It contains reference to either the `byte[]` of the stack trace (when there are no dynamic methods on the stack trace) or an `object[]` where the first element contains the stack trace `byte[]` reference and the following elements contain what used to be in the dynamic array. That allows atomic updates and reads of the stack trace and dynamic method keepalive references without a need of a lock.

I have also removed a dead code path from the `StackTraceInfo::SaveStackTrace`.

janvorli committed:
Remove global spinlock for EH stacktrace

The global spinlock that was used to ensure that stack trace and the
associated dynamic methods array were updated and read atomically.
However, for the new EH, it has shown to cause a high contention in case
many threads were handling exceptions at the same time.

This change replaces the two arrays by one object member in the
exception class. It contains reference to either the byte[] of the
stack trace (when there are no dynamic methods on the stack trace) or an
object[] where the first element contains the stack trace byte[]
reference and the following elements contain what used to be in the
dynamic array. That allows atomic updates and reads of the stack trace
and dynamic method keepalive references without a need of a lock.

I have also removed a dead code path from the
StackTraceInfo::SaveStackTrace.

janvorli commented:
/azp run runtime-coreclr outerloop

janvorli commented:
/azp run runtime-coreclr gcstress0x3-gcstress0xc

jkotas commented:
Tests are crashing:
```
023ac164 7202552a coreclr!_DbgBreakCheckNoThrow+0x51
023ac200 71ae5c5f coreclr!DbgAssertDialog+0x20c
023ac218 71b692af coreclr!ArrayBase::GetDataPtrOffset+0x29
(Inline) -------- coreclr!ArrayBase::GetDataPtr+0x12
023ac224 71c55e00 coreclr!Array<signed char>::GetDirectPointerToNonObjectElements+0x15
(Inline) -------- coreclr!StackTraceArray::GetHeader+0x5
(Inline) -------- coreclr!StackTraceArray::GetObjectThread+0x5
023ac264 71c554f3 coreclr!StackTraceArray::EnsureThreadAffinity+0x19
023ac2f4 71ba4b9a coreclr!StackTraceArray::Append+0x6b
023ac4c4 71f664b5 coreclr!StackTraceInfo::SaveStackTrace+0x3fd
023ac598 71c9918b coreclr!COMPlusThrowCallback+0x285
```

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Exception.CoreCLR.cs:
```diff
@@ -181,7 +194,6 @@ internal void RestoreDispatchState(in DispatchState dispatchState)
         // DynamicMethodDescs alive for the lifetime of the exception. We do this because
```
I assume that we use this array to also store loader allocator roots when collectible assemblies are referenced in the stacktrace. Is that correct? The names and comments that talk about DynamicMethods are misleading in that case.

janvorli replied:
Yes, we do. I'll update the related comments.

janvorli replied:
Actually, the renaming is very pervasive, let's make it as a separate change to not to cause more churn for the review.

jkotas commented:
> without a need of a lock.

I do not see how you can get away without a lock. The stacktrace attached to the exception can be accessed by getting modified by multiple thread simultaneously when the same exception instance is re-thrown on multiple threads in parallel.

janvorli committed:
Fix few issues and realloc dynamic methods array for non-creator threads

janvorli commented:
/azp run runtime-coreclr outerloop

janvorli commented:
/azp run runtime-coreclr gcstress0x3-gcstress0xc

jkotas commented:
This breaks stacktraces for re-thrown exceptions. I am surprised that it does not cause any test faulures. Could you please make add a test that would catch this regression?

Repro: https://gist.github.com/jkotas/53c941dc7408ea53a5c11dbc9f0a9fea

## Expected output
```
System.Exception: My exception
   at Program.ThrowHelper()
   at MyMethod4()
   at Program.AppendDynamicFrame()
   at Program.ThrowHelper()
   at MyMethod3()
   at Program.AppendDynamicFrame()
   at Program.ThrowHelper()
   at MyMethod2()
   at Program.AppendDynamicFrame()
   at Program.ThrowHelper()
   at MyMethod1()
   at Program.AppendDynamicFrame()
   at Program.ThrowHelper()
   at MyMethod0()
   at Program.AppendDynamicFrame()
   at Program.ThrowHelper()
   at Program.Main()
```

## Actual output
```
System.Exception: My exception
   at Program.ThrowHelper()
   at MyMethod4()
```

Once you fix the functional issue, could you please double-check for race conditions in the presence of other threads throwing the exception, formatting the stacktrace, etc. (I suspected that there are some race conditions and I wanted to demonstrate them using the test, only to find that there are more fundamental functional issues.)

janvorli commented:
@jkotas thank you for finding that bug, it also surprises me that none of our tests fails due to that. I have also already spent some time reasoning about possible races, I have found one issue that I've introduced and have a local fix. But overall, the original code was using the lock only to read / write the stack trace and dynamic methods arrays atomically. It already had a measure to create a copy of the stack trace array in case the thread that was adding to it was different from the one that has created it. That's inside the StackFrameArray class implementation and I have not changed it. The dynamic methods array access was not doing anything special in case of multi-threaded race and I think that was possibly creating some race conditions that would matter. That's why in my recent commit, I've also done the same thing for the array of resolvers / loader allocators. I've made on mistake in it that causes one of the tests to crash with GC stress, which I have fixed locally already. 
I don't currently see a possibility of race in the current state, but I am definitely planning to take some more time reasoning about it. The code of the SaveStackTrace is very convoluted and kind of difficult to reason about, I would like to clean it up later to make it easier to read.

janvorli committed:
Fix two issues

* missing stack trace parts in exception rethrow
* missing copying of the resolver / loader allocator array when a new
  thread cloned the stack trace array, but no new collectible methods
  were added

janvorli commented:
@jkotas the last commit has fixed the issue with stack trace you were seeing and also fixed a race when the dynamic methods array was not copied when the stack trace was (the coreclr\GitHub_45929\test45929 test was failing due to that with GC stress enabled). So if you want to give it a try, it is ready for that.
Other than that, I still plan to add the test you have asked for to this PR.

janvorli commented:
/azp run runtime-coreclr gcstress0x3-gcstress0xc

janvorli commented:
/azp run runtime-coreclr outerloop

jkotas commented on src/coreclr/vm/excep.cpp:
```diff
@@ -2460,7 +2455,15 @@ void StackTraceInfo::SaveStackTrace(BOOL bAllowAllocMem, OBJECTHANDLE hThrowable
-                ((EXCEPTIONREF)ObjectFromHandle(hThrowable))->SetStackTrace(gc.stackTrace.Get(), gc.dynamicMethodsArray);
+                if (gc.dynamicMethodsArray != NULL)
+                {
+                    gc.dynamicMethodsArray->SetAt(0, gc.stackTrace.Get());
```
The stacktrace is visible to other threads at this point. It is too late to set the dynamicMethodArray here.

Here is a repro for this race condition:
1. Add sleep in front of this line to make a deterministic simple repro possible:
```diff
                 if (gc.dynamicMethodsArray != NULL)
                 {
+                    ClrSleepEx(2000,FALSE);
+
                     gc.dynamicMethodsArray->SetAt(0, gc.stackTrace.Get());
                     ((EXCEPTIONREF)ObjectFromHandle(hThrowable))->SetStackTrace(dac_cast<OBJECTREF>(gc.dynamicMethodsArray));
                 }
```
2. Run https://gist.github.com/jkotas/12fdbd933a1ad0c1cc32a96353c4b1d4 (on checked build)
3. You should see it to fail with:
```
Assert failure(PID 20804 [0x00005144], Thread: 13800 [0x35e8]): handle

CORECLR! LCGMethodResolver::GetManagedResolver + 0x2C (0x00007ff8`4f3f1aac)
CORECLR! DebugStackTrace::GetStackFramesInternal + 0xF80 (0x00007ff8`4f3dd3f0)
```

jkotas reviewed (changes requested):
It would be useful to add a comment that explains the thread safety guarantees for the captured stacktraces.

jkotas commented:
I think there needs to be some sort of lock to protect appending to the existing stacktrace. It can be a spin lock.

janvorli commented on src/coreclr/vm/object.cpp:
```diff
@@ -1494,7 +1494,7 @@ void StackTraceArray::EnsureThreadAffinity()
+            this->Set(copy.Get());
         GCPROTECT_END();
     }
 }
```
The `StackTraceArray` has a header (`StackTraceArray::ArrayHeader`) containing a m_size member that's updated as the last item when the array is modified (see `StackTraceArray::Append`). Since readers only read upto "size" number of elements, there should be no torn reads possible.

jkotas replied:
This works as long as the array is not reused and overwritten.

I see comments like https://github.com/dotnet/runtime/pull/100427/files#diff-47cbfa47fb88845e5e76fd4df0a84e9da494508ffe80da116511fc2d0becd9baR2248 that make it sound like that there are situations where the array is reused, but they just may be stale or misleading.

janvorli replied:
It is never reused or overwritten. It is either appended to or replaced by a brand new one - either by an empty one or by a new one with a copy of the contents of the old one, but only the `m_size` elements are copied. 
I wonder which of the comments you are referring to, the link has shown me just the whole diff without anything specific marked.

jkotas replied:
> I wonder which of the comments you are referring to

```
// Cleanup previous info
gc.stackTrace.Append(m_pStackTrace, m_pStackTrace + m_dFrameCount);
```

janvorli replied:
Ah, I can see how this is misleading. The gc.StackTrace is empty at this point (the m_array is NULL), the Append results in allocating a new managed array and copying the frames into it. And in the end, the gc.stackTrace is written to the Exception, essentially clearing what was there before and creating a new stack trace from scratch.

leculver commented on src/coreclr/debug/daccess/enummem.cpp:
```diff
@@ -533,7 +533,7 @@ HRESULT ClrDataAccess::DumpManagedExcepObject(CLRDataEnumMemoryFlags flags, OBJE
 
     // Dump the stack trace array object and its underlying type
-    I1ARRAYREF stackTraceArrayObj = exceptRef->GetStackTraceArrayObject();
+    OBJECTREF stackTraceArrayObj = exceptRef->GetStackTraceArrayObject();
```
When capturing a minidump or triage dump, we need to ensure that the underlying stack trace objects make their way into the dump file.  The code in enummem.cpp only dump the first level object, which is now just the `object[]` which contains the pointer to the `byte[]` we care about.

So I think this logic needs to be expanded so that everything within `if (stackTraceArrayObj != NULL)` (following the above line) is duplicated for the first array element in stackTraceArrayObj.

I don't think we need to capture the extra entries in the object[] for a minidump, but if those entries would be useful in a minidump scenario, please consider adding those too.

janvorli replied:
@leculver thank you for spotting this. I will make sure the `byte[]` array from the `object[]` is dumped too. As for dumping the rest of the `object[]` for minidump, we were not dumping the `_dynamicMethods` array before either (that was where those entries were before my change) and I also don't see them to be useful in minidumps.

janvorli commented:
I am closing this PR and creating a new one that supersedes it.
